build-system/tasks/changelog.js

```diff
@@ -0,0 +1,169 @@
+var argv = require('minimist')(process.argv.slice(2));
+var config = require('../config');
+var extend = require('util')._extend;
+var fs = require('fs');
+var git = require('gulp-git');
+var gulp = require('gulp');
+var request = BBPromise.promisify(require('request'));
+var util = require('gulp-util');
+
+var GITHUB_ACCESS_TOKEN = process.env.GITHUB_ACCESS_TOKEN;
+var gitExec = BBPromise.promisify(git.exec);
+
+
+function changelog() {
+  if (!GITHUB_ACCESS_TOKEN) {
+    util.log(util.colors.red('Warning! You have not set the ' +
+        'GITHUB_ACCESS_TOKEN env var. This task might hit the default ' +
+        'rate limit set by github (60).'));
+    util.log(util.colors.green('See https://help.github.com/articles/' +
+        'creating-an-access-token-for-command-line-use/ ' +
+        'for instructions on how to create a github access token. We only ' +
+        'need `public_repo` scope.'));
+  }
+
+  return getGitMetadata();
+}
+
+function getGitMetadata() {
+  var gitMetadata = {};
+  return getGitTag(gitMetadata)
+      .then(onGitTagSuccess.bind(null, gitMetadata))
+      .then(getGitLog)
+      .then(onGitLogSuccess.bind(null, gitMetadata))
+      .then(fetchGithubMetadata)
+      .then(buildChangelog.bind(null, gitMetadata))
+      .then(function(gitMetadata) {
+        // output to the console so that an external script
+        // can read the information.
+        console/*OK*/.log(gitMetadata.changelog);
+      })
+      .catch(errHandler);
+}
+
+function buildChangelog(gitMetadata, githubMetadata) {
+  var titles = githubMetadata
+      .filter(function(data) {
+        return !data.filenames.every(function(filename) {
+          return config.changelogIgnoreFileTypes.test(filename);
+        });
+      })
+      .map(function(data) {
+        return '  - ' + data.title.trim();
+      }).join('\n');
+  gitMetadata.changelog = titles;
+  return gitMetadata;
+}
+
+function getGitTag() {
+  var options = {
+    args: 'describe --abbrev=0 --tags'
+  };
+  return gitExec(options);
+}
+
+function getGitLog(tag) {
+  var options = {
+    args: 'log master...' + tag + ' --pretty=format:%s --merges'
+  };
+  return gitExec(options);
+}
+
+function fetchGithubMetadata(ids) {
+  var options = {
+    url: 'https://api.github.com/repos/ampproject/amphtml/pulls/',
+    headers: {
+      'User-Agent': 'amp-changelog-gulp-task'
+    }
+  };
+
+  if (GITHUB_ACCESS_TOKEN) {
+    options.qs = {
+      access_token: GITHUB_ACCESS_TOKEN
+    }
+  }
+
+  // NOTE: (erwinm) not sure if theres a better way to do this, we're
+  // doing n + 1 fetches here since we can't batch things up.
+  var requests = ids.map(function(id) {
+    var prOption = extend({}, options);
+    prOption.url += id;
+
+    return getPullRequestTitle(prOption).then(function(title) {
+      var filesOption = extend({}, prOption);
+      filesOption.url += '/files';
+      return getPullRequestFiles(title, filesOption);
+    });
+  });
+
+  return BBPromise/*OK*/.all(requests);
```

Comment:
Yep, just went for consistency here since i was using bluebird's `Promise.promisify` (converts node's callback convention into promisified functions). I can change if you feel thats more appropriate.